### DISCUSSION TITLE
Speeding up `CRootOf.n()`

### DISCUSSION THREAD

**Discussion by skieffer** [sympy/sympy#25193] [Q&A]
I was thinking about trying to work on @oscarbenjamin 's  [idea](https://github.com/sympy/sympy/pull/22737#issuecomment-1000592349) for an interval arithmetic complex Newton iteration, in hopes of speeding up `CRootOf.n()`. But I realized that I first have a question: `CRootOf` already has *two* built in methods for computing a floating point approximation: one fast and one slow. It uses only the slow one. Why?

## Review

When you ask for a floating point approximation of a complex root, you're going to invoke `CRootOf._eval_evalf()`. That currently looks like this:

https://github.com/sympy/sympy/blob/a76b02fcd3a8b7f79b3a88df50c19eb7aee33c17/sympy/polys/rootoftools.py#L919-L922

Thus, it simply uses `CRootOf.eval_rational()`.

Meanwhile, `CRootOf` has a second method that could have been used instead of `eval_rational()`, called `eval_approx()`.

## Is `eval_approx()` faster than `eval_rational()`?

Yes, by a factor of about 100 to 150 times. I ran them both on random polys, of various degrees and coeff bounds.

<details>
<summary>CLICK TO SEE TIMING RESULTS</summary>

```
--------------------------------------------------------------------------------
deg: 3, num polys: 10, coeff bound: 10, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   5.76087 5.80314 5.82405 5.88500 6.09751 6.11714 6.13601 6.22713 6.41657 6.57491
   AVG 6.08423
approx
   0.02446 0.02522 0.02545 0.02879 0.03157 0.03299 0.03581 0.03689 0.04639 0.23216
   AVG 0.05197
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 117.07
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
deg: 4, num polys: 10, coeff bound: 10, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   6.12078 6.18026 7.12837 7.56812 8.06391 8.52039 8.60004 8.81123 8.98332 9.24432
   AVG 7.92207
approx
   0.02548 0.02980 0.04072 0.04158 0.04427 0.05338 0.05462 0.05558 0.05645 0.05841
   AVG 0.04603
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 172.11
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
deg: 5, num polys: 10, coeff bound: 10, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   7.64363 10.57060 11.35892 11.40954 11.54209 11.86983 12.01539 12.54192 13.25539 13.36179
   AVG 11.55691
approx
   0.04616 0.05834 0.06369 0.06842 0.07092 0.07130 0.07519 0.08817 0.10169 0.12492
   AVG 0.07688
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 150.32
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
deg: 6, num polys: 10, coeff bound: 10, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   11.74742 12.53840 14.84450 15.51244 15.66906 16.05221 17.53238 17.54438 17.91116 18.00991
   AVG 15.73619
approx
   0.04348 0.07267 0.08190 0.10496 0.11527 0.11624 0.15651 0.15659 0.15728 0.29807
   AVG 0.13030
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 120.77
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
deg: 7, num polys: 10, coeff bound: 10, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   16.75855 20.12493 24.03068 24.23696 24.27473 24.68701 24.96107 25.02804 25.13900 26.81829
   AVG 23.60593
approx
   0.10753 0.13810 0.14219 0.15198 0.16820 0.17586 0.20373 0.20611 0.24226 0.26611
   AVG 0.18021
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 130.99
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
deg: 8, num polys: 10, coeff bound: 10, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   25.06398 29.41533 29.87320 31.42546 33.92145 34.55257 35.84201 36.70357 40.41141 41.91209
   AVG 33.91211
approx
   0.10970 0.16933 0.17474 0.19200 0.21124 0.23037 0.25234 0.25868 0.26424 0.34426
   AVG 0.22069
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 153.66
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
Now with 100 as coeff bound. Seems basically the same...
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
deg: 3, num polys: 10, coeff bound: 100, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   5.79767 6.13475 6.14930 6.15773 6.29940 6.36568 6.41947 6.59281 6.63194 6.66550
   AVG 6.32142
approx
   0.02944 0.03124 0.03270 0.03497 0.03976 0.04135 0.04544 0.06652 0.06941 0.28383
   AVG 0.06747
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 93.70
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
deg: 4, num polys: 10, coeff bound: 100, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   7.57426 7.62958 8.05890 8.08199 8.34423 8.48957 9.15493 9.37372 9.49552 9.99107
   AVG 8.61938
approx
   0.02720 0.03335 0.03851 0.04181 0.04386 0.04517 0.06601 0.08903 0.08952 0.09682
   AVG 0.05713
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 150.88
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
deg: 5, num polys: 10, coeff bound: 100, dps: 100
--------------------------------------------------------------------------------
TIME (s)
rational
   10.99970 11.70508 11.75977 11.83307 11.86854 11.96035 11.96894 12.58834 12.62541 12.65391
   AVG 11.99631
approx
   0.05869 0.06685 0.06930 0.07555 0.09526 0.09789 0.11308 0.11480 0.14380 0.14458
   AVG 0.09798
--------------------------------------------------------------------------------
RATIO
  AVG rational / AVG approx: 122.44
--------------------------------------------------------------------------------
```

</details>

## Is `eval_approx()` a suitable replacement for `eval_rational()`?

I tried changing the implementation of `CRootOf._eval_evalf()` so that it would use `eval_approx()` instead of `eval_rational()`:

```python
    def _eval_evalf(self, prec, **kwargs):
        """Evaluate this complex root to the given precision."""
        # all kwargs are ignored
        return self.eval_approx(prec_to_dps(prec))
```

and ran all the unit tests. They all passed.

So what could be the reason for using `eval_rational()`? Does it have some important side effects? The only side effect I can think of is the refinement of the isolating intervals for the roots, which are cached, and do represent some work that's worth saving. But `eval_approx()` refines the intervals and populates the cache too:

```python
from sympy import CRootOf, Poly
from sympy.abc import x
import sympy.polys.rootoftools as rootoftools

f = Poly(5*x**3 + 3*x**2 + 6*x + 7, x, domain='ZZ')
r = CRootOf(f, -1)
assert not r.is_real

CRootOf.clear_cache()
rootoftools._complexes_cache._dict
Out[18]: {}

r.eval_approx(100)
rootoftools._complexes_cache._dict
Out[20]: 
{PurePoly(5*x**3 + 3*x**2 + 6*x + 7, x, domain='ZZ'): [(7/80, 7/40) x (-49/40, -91/80),
  (7/80, 7/40) x (91/80, 49/40)]}
```

It happens right here:

https://github.com/sympy/sympy/blob/a76b02fcd3a8b7f79b3a88df50c19eb7aee33c17/sympy/polys/rootoftools.py#L911-L913

**Comment by oscarbenjamin** [comment]
I just took a quick look at `eval_approx` and it is not clear to me that it is in any sense formally verifiable. Also:
https://github.com/sympy/sympy/blob/a76b02fcd3a8b7f79b3a88df50c19eb7aee33c17/sympy/polys/rootoftools.py#L907
Why is this catching UnboundLocalError?

The key to the interval arithmetic approach that I suggested is that it uses approximate arithmetic but in a way that is still provably correct through outer rounding. The outer rounding means that in each arithmetic operation the lower bound rounds down and the upper bound rounds up. That way we can say for sure that the result of some calculation is definitely between the upper and lower bounds that we end up with. Ultimately what means that although there is approximate arithmetic and rounding we can still use that to say "this interval definitely contains exactly one root".

**Comment by skieffer** [comment]
So, I don't know a lot about these kinds of root finding methods, but my basic impression is that the "core, iterative processes" like the `findroot()` method of `mpmath` that's being used in `eval_approx()`, are sort of error prone, in that they can do things like (a) get stuck in a local extremum, or (b) converge to a root other than the one we were intersted in.

But it looks to me like `eval_approx()` has been designed to try to handle all those sorts of issues:

https://github.com/sympy/sympy/blob/a76b02fcd3a8b7f79b3a88df50c19eb7aee33c17/sympy/polys/rootoftools.py#L834-L837

If you look carefully at the comment inside the `try` block, the `UnboundLocalError` is evidently something `findroot()` might raise, when a particular thing goes wrong:

https://github.com/sympy/sympy/blob/a76b02fcd3a8b7f79b3a88df50c19eb7aee33c17/sympy/polys/rootoftools.py#L895-L898

Apparently refining the interval and trying again is supposed to be a good enough response.

When you say "formally verifiable," are you basically saying that we're lacking a proof that `eval_approx()` always works? In other words, without a literature citation, or a theorem to point to, we can't responsibly offer this method as a root finder. Is that what you mean?

**Reply by oscarbenjamin** [reply]
> When you say "formally verifiable," are you basically saying that we're lacking a proof that `eval_approx()` always works? In other words, without a literature citation, or a theorem to point to, we can't responsibly offer this method as a root finder. Is that what you mean?

No, what I mean is that I am sure that eval_approx is based on heuristic methods that are not guaranteed to be correct (i.e. `find_root`). The interval method that I showed is not heuristic in any way. The idea of interval Newton is the following:

If we start with an interval $I_i$ containing a point $p$ and we compute an interval $I_{i+1}$ as
$$I_{i+1} = p - \frac{f(p)}{f'(I_i)}$$
Now any root of `f` that lay in $I_i$ must also lie in $I_{i+1}$ (so we use their intersection going forwards). This iteration typically shrinks the interval but if not just split the interval in two and iterate on each interval separately instead. The big theorem here is this:

> If $f'(I_i)$ does  not contain zero and $I_{i+1}$ is a strict subset of $I_i$ then it is proven that $f$ has a unique root in $I_i$.

The key feature of interval Newton is that we can use it to *prove* that a given interval contains a unique root. In the case of irreducible polynomials we already know exactly how many roots there are and we know that $f'(x)$ is nonzero at every root. What this means is that any method can be used to find the roots: we can prove their bounding intervals correct after the fact. This means that an approach like `eval_approx` can be used as long as we have a final step of interval Newton at the end to prove correctness.

I expect that the current root refinement code already uses some variant of interval Newton but it does so using exact rational arithmetic which is inefficient. We do not need infinitely exact arithmetic, just high precision arithmetic with careful rounding. The theorem that proves `$I_i$ has a unique root still applies under approximate interval arithmetic as long as the end points of the intervals always round outwards.
